tests/test_inference.py

```diff
@@ -2508,6 +2518,228 @@ def test_many_trees(self):
+    tables.nodes.add_row(time=2)
+    tables.edges.add_row(parent=1, child=0, left=0, right=1)  # "dead" topology
+    yield tables.tree_sequence(), "Topology but no samples"
+
+    first_sample = tables.nodes.add_row(flags=tskit.NODE_IS_SAMPLE)
+    yield tables.tree_sequence(), "Isolated_node_1"
+
+    # Make a few isolated nodes
+    tables.nodes.add_row(flags=tskit.NODE_IS_SAMPLE)
+    tables.nodes.add_row(flags=tskit.NODE_IS_SAMPLE)
+    ts = tables.tree_sequence()
+    assert ts.first().num_roots > 1
+    yield ts, "Isolated_nodes_3"
+
+    tables.edges.add_row(parent=0, child=first_sample, left=0, right=1)
+    tables.sort()
+    ts = tables.tree_sequence()
+    assert ts.first().num_roots > 1
+    yield ts, "Partial_3_root"
+
+    tables.nodes.truncate(first_sample + 1)
+    ts = tables.tree_sequence()
+    assert ts.num_samples == 1
+    yield ts, "Partial_1_root"
+
+    tables.edges.add_row(parent=1, child=first_sample, left=1, right=ts.sequence_length)
+    tables.sort()
+    ts = tables.tree_sequence()
+    assert ts.at_index(0).root == ts.at_index(1).root
+    yield ts, "Two_edges_to_1_root"
+
+    tables.nodes.add_row(flags=tskit.NODE_IS_SAMPLE)
+    ts = tables.tree_sequence()
+    assert ts.at_index(0).num_roots > 1
+    yield ts, "Two_edges_to_1_root_plus_isolated"
+
+    tables.nodes.truncate(first_sample + 1)
+    tables.edges.truncate(len(tables.edges) - 1)
+    tables.edges.add_row(parent=0, child=first_sample, left=1, right=ts.sequence_length)
+    tables.sort()
+    ts = tables.tree_sequence()
+    assert ts.at_index(0).root != ts.at_index(1).root
+    yield ts, "Two_edges_to_2_roots"
+
+    oldest = tables.nodes.add_row(time=3)
+    tables.edges.add_row(parent=oldest, child=1, left=0, right=ts.sequence_length)
+    u = tables.nodes.add_row(flags=tskit.NODE_IS_SAMPLE)  # add an isolated node too
+    yield tables.tree_sequence(), "Single_root_edge"
+
+    tables.edges.add_row(parent=oldest, child=u, left=0, right=ts.sequence_length)
+    yield tables.tree_sequence(), "Multi_root_edge"
+
+
+def ts_root_test_params():
+    return [pytest.param(ts, id=name) for ts, name in ts_root_tests()]
+
+
+class TestRootCheckingFunctions:
+    """
+    Tests some helper functions such as has_single_edge_over_grand_root and
+    has_same_root_everywhere
+    """
+
+    @pytest.mark.parametrize("ts", ts_root_test_params())
+    def test_has_same_root_everywhere(self, ts):
+        roots = {root for tree in ts.trees() for root in tree.roots}
+        if len(roots) == 1 and tskit.NULL not in roots:
+            assert tsinfer.has_same_root_everywhere(ts)
+        else:
+            assert not tsinfer.has_same_root_everywhere(ts)
+
+    @pytest.mark.parametrize("ts", ts_root_test_params())
+    def test_has_single_edge_over_grand_root(self, ts, request):
+        if request.node.callspec.id == "Single_root_edge":
+            assert tsinfer.has_single_edge_over_grand_root(ts)
```

Comment:
This is brittle and needs the reader to understand the innards of pytest. Better to break up the list of examples (or to make this property a parameter of ``ts_root_test_params()``?)